src/ChannelManager.cpp

```diff
@@ -0,0 +1,210 @@
+// ChannelManager.cpp
+
+// Implements the cChannelManager class which stores and calls channel callbacks
+
+#include "Globals.h"  // NOTE: MSVC stupidness requires this to be the same across all modules
+
+#include <algorithm>
+
+#include "ChannelManager.h"
+#include "ChannelCallback.h"
+#include "ClientHandle.h"
+#include "Entities/Player.h"
+#include "OSSupport/CriticalSection.h"
+#include "World.h"
+
+
+class cChannelBroadcastCallback : public cPlayerListCallback
+{
+	AString m_Channel;
+	AString m_Data;
+public:
+	cChannelBroadcastCallback(const AString & a_Channel, const AString & a_Data)
+	: m_Channel(a_Channel), m_Data(a_Data){}
+
+	virtual bool Item(cPlayer * a_Type)
+	{
+		cClientHandle * Handle = a_Type->GetClientHandle();
+		if (Handle->HasPluginChannel(m_Channel))
+		{
+			Handle->SendPluginMessage(m_Channel, m_Data);
+		}
+		return false;
+	}
+};
+
+
+
+
+
+bool cChannelManager::RegisterChannel(const AString & a_Channel, cChannelCallbackPtr a_Callback)
+{
+	cCSLock Lock(m_CSCallbacks);
+	if (m_ChannelCallbacks.find(a_Channel) == m_ChannelCallbacks.end())
+	{
+		m_ChannelCallbacks.insert(std::make_pair(a_Channel, a_Callback));
+		return true;
+	}
+
+	return false;
+}
+
+
+
+
+
+bool cChannelManager::RemoveChannel(const AString & a_Channel)
+{
+	cCSLock Lock(m_CSCallbacks);
+	auto callback = m_ChannelCallbacks.find(a_Channel);
+	if (callback != m_ChannelCallbacks.end())
+	{
+		m_ChannelCallbacks.erase(callback);
+		return true;
+	}
+	return false;
+}
+
+
+
+
+
+void cChannelManager::AddClientToChannel(cClientHandle & a_Client, const AString & a_Channel)
+{
+	a_Client.RegisterChannel(a_Channel);
+	a_Client.SendPluginMessage("REGISTER", a_Channel);
+}
+
+
+
+
+
+void cChannelManager::AddClientToChannels(cClientHandle & a_Client, const AString & a_Channels)
+{
+	auto channels = BreakApartChannels(a_Channels);
+	a_Client.RegisterChannels(channels);
+}
+
+
+
+
+
+void cChannelManager::RemoveClientFromChannel(cClientHandle & a_Client, const AString & a_Channel)
+{
+	a_Client.UnregisterChannel(a_Channel);
+	a_Client.SendPluginMessage("UNREGISTER", a_Channel);
+}
+
+
+
+
+
+void cChannelManager::RemoveClientFromChannels(cClientHandle & a_Client, const AString & a_Channels)
+{
+	auto channels = BreakApartChannels(a_Channels);
+	a_Client.UnregisterChannels(channels);
+}
+
+
+
+
+
+bool cChannelManager::ClientHasChannel(cClientHandle & a_Client, const AString & a_Channel)
```

Comment:
Okay.

> However, someone was working on moving cClientHandles to being managed by the server
> 
Sounds like #3115 :D


On 4 Aug 2016, at 02:14, "Samuel Barney" <notifications@github.com> wrote:

> In src/ChannelManager.cpp:
> 
> > +
> > +
> > +
> > +
> > +
> > +void cChannelManager::RemoveClientFromChannels(cClientHandle & a_Client, const AString & a_Channels)
> > +{
> > +	auto channels = BreakApartChannels(a_Channels);
> > +	a_Client.UnregisterChannels(channels);
> > +}
> > +
> > +
> > +
> > +
> > +
> > +bool cChannelManager::ClientHasChannel(cClientHandle & a_Client, const AString & a_Channel)
> Well, the majority of this code was written to be a halfway point. Right now it just merely wraps code in cClientHandle. However, someone was working on moving cClientHandles to being managed by the server, and I wanted to move all channel related code and data into cChannelManager. I don't know if that move has been accomplished or not, so I wanted to leave it at this halfway point until I know for certain.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>